Sources/AsyncHTTPClient/ConnectionPool/State Machine/HTTPConnectionPool+HTTP2Connections.swift

```diff
@@ -0,0 +1,532 @@
+            guard let index = self.connections.firstIndex(where: { $0.connectionID == connectionID }) else {
+                preconditionFailure("go away recieved for a connection that does not exists")
+            }
+            self.connections[index].goAwayReceived()
+        }
+
+        mutating func newHTTP2MaxConcurrentStreamsReceived(
+            _ connectionID: Connection.ID,
+            newMaxStreams: Int
+        ) -> (Int, AvailableConnectionContext) {
+            guard let index = self.connections.firstIndex(where: { $0.connectionID == connectionID }) else {
+                preconditionFailure("We tried to update the maximum concurren streams number for a connection that does not exists")
+            }
+            let availableStreams = self.connections[index].newMaxConcurrentStreams(newMaxStreams)
+            let context = AvailableConnectionContext(
+                availableStreams: availableStreams,
+                eventLoop: self.connections[index].eventLoop,
+                isIdle: self.connections[index].isIdle
+            )
+            return (index, context)
+        }
+
+        // MARK: Leasing and releasing
+
+        mutating func leaseStream(onPreferred eventLoop: EventLoop) -> Connection? {
+            guard let index = self.findAvailableConnection(onPreferred: eventLoop, minimumAvailableStreams: 1) else { return nil }
+            return self.leaseStreams(at: index, count: 1)
+        }
+
+        /// tries to find an available connection on the prefered `eventLoop`. If it can't find one with the given `eventLoop`, it returns the first available connection
+        private func findAvailableConnection(onPreferred eventLoop: EventLoop, minimumAvailableStreams: Int) -> Int? {
+            var availableConnection: Int?
+            for (index, connection) in self.connections.enumerated() {
+                guard connection.canLeaseStreams(count: minimumAvailableStreams) else { continue }
+                if connection.eventLoop === eventLoop {
+                    return index
+                } else if availableConnection == nil {
+                    availableConnection = index
+                }
+            }
+            return availableConnection
+        }
+        
+        mutating func leaseStreams(onRequired eventLoop: EventLoop, count: Int) -> Connection? {
+            guard let index = self.findAvailableConnection(onRequired: eventLoop, minimumAvailableStreams: count) else { return nil }
+            return self.leaseStreams(at: index, count: count)
+        }
+
+        /// tries to find an available connection on the required `eventLoop`
+        private func findAvailableConnection(onRequired eventLoop: EventLoop, minimumAvailableStreams: Int) -> Int? {
+            self.connections.firstIndex(where: { $0.eventLoop === eventLoop && $0.canLeaseStreams(count: minimumAvailableStreams) })
+        }
+
+        /// - precondition: `index` needs to be valid
```

Comment:
We should give a note where to get the index from. index was received from a call to this or that with a `AvailableConnectionContext`.